src/time_steppers.jl

```diff
@@ -354,11 +367,16 @@ apply_bcs!(arch, ::Val{:z}, Bx, By, Bz, args...) =
 "Apply a top and/or bottom boundary condition to variable ϕ. Note that this kernel
 MUST be launched with blocks=(Bx, By). If launched with blocks=(Bx, By, Bz), the
 boundary condition will be applied Bz times!"
-function apply_z_bcs!(top_bc, bottom_bc, grid, ϕ, Gϕ, κ, t, iteration, u, v, w, T, S)
+function apply_z_bcs!(top_bc, bottom_bc, grid, ϕ, Gϕ, κ, closure, eos, g, t, iteration, u, v, w, T, S)
     @loop for j in (1:grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
         @loop for i in (1:grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
-            apply_z_top_bc!(top_bc, i, j, grid, ϕ, Gϕ, κ, t, iteration, u, v, w, T, S)
-            apply_z_bottom_bc!(bottom_bc, i, j, grid, ϕ, Gϕ, κ, t, iteration, u, v, w, T, S)
+
+               κ_top = κ(i, j, 1,       grid, closure, eos, g, u, v, w, T, S)
+            κ_bottom = κ(i, j, grid.Nz, grid, closure, eos, g, u, v, w, T, S)
+
+               apply_z_top_bc!(top_bc,    i, j, grid, ϕ, Gϕ, κ_top,    t, iteration, u, v, w, T, S)
+            apply_z_bottom_bc!(bottom_bc, i, j, grid, ϕ, Gϕ, κ_bottom, t, iteration, u, v, w, T, S)
```

Comment:
It looks like if you want to impose a `z` boundary condition that does not depend on `κ`, you still have to calculate κ using the full closure which can be expensive if using an LES closure. Not sure how to get around this, probably some clever multiple dispatch?

This is probably fine for now as constant Smagorinsky isn't integrated yet, and the performance hit probably isn't big enough to worry about right now.

Reply:
Hmm yes that is possible. It doesn't have to clever :-P --- we will just pass the function to calculate the transport coefficient into `apply_z_top_bc!`, which dispatches on the type of its first argument (the boundary condition now). That's possibly a better design anyways.

But actually, it might get compiled away if not used anyways --- the type of both boundary conditions is known to the function `apply_z_bcs!`.

I didn't put a significant amount of effort into this problem because I'm expecting a lot to simplify once we have halo regions.